java/src/com/google/idea/blaze/java/sync/model/BlazeJarLibrary.java

```diff
@@ -66,7 +66,7 @@ public void modifyLibraryModel(
     libraryModel.addRoot(pathToUrl(jar), OrderRootType.CLASSES);
 
     AttachedSourceJarManager sourceJarManager = AttachedSourceJarManager.getInstance(project);
-    if (!sourceJarManager.hasSourceJarAttached(key)) {
```

Comment:
Thanks @brendandouglas! I understand what you're saying. when I saw that code, it looked weird to me, perhaps because I don't have a good picture of how the model is populated exactly. However, libraries are never populated with source jars, so I still need to find the problem, if that's not it. 

I also noticed changes in the way `LibraryArtifact` exposes information about library jars that might be relevant. Specifically the removal of method `getSourceJarsOrClassJar` and the changes in `jarForIntellijLibrary`. So I'll take a look in that area as well. 

In any case, I'll appreciate any leads you can give me, that might make my life easier.

Thank you!

Reply:
Source jars aren't attached on sync, for performance reasons (it's expensive to index all source jars for larger projects).

When you first navigate to a class file in a library with a source jar, the source jar should then be attached (and you should be redirected to the java file in the source jar).

I haven't been able to reproduce a situation where that doesn't happen. If you could point me to a project where it's not working as expected, that would help track it down.

Reply:
Thanks! Now I know how it is expected to work, so that might give me pointers.. 
I'm working on Scala support mainly, so it might be related to that as well. The libraries information is provided by `rules_scala` seems to be correct though.

I have one very small project in which I actually have it working and several pretty big ones in which it doesn't work. In the projects that don't load sources, the XML files created under `.ijwb/.idea/libraries` do not have the `SOURCES` populated at all - that's why I thought that a call to populate the library model was completely missing. I will dig more into it on Sunday and hopefully I will be able to give you an example. If you have any tip for me regarding where to start debugging, it will be great.

Thanks again.

Reply:
@brendandouglas I created [that](https://github.com/sha1n/bazel_scala_sample) repository as an example. As I mentioned before, it's a Scala project, so you need to add `scala` to `additional_languages` in the project file.

After you import, go to `com.example.scala.Example` and try to click thorough to `ClientTransport`, which should load its source.

From my last debugging session it seems that source-jar navigation is only implemented for Java (`BlazeSourceJarNavigationPolicy`) and will not work for libraries imported by `BlazeScalaWorkspaceImporter` because the code that's looking for source jars looks specifically at `BlazeJavaSyncData` and is rightfully unaware of `BlazeScalaSyncData`. The question is what is the correct way to implement it. Should the Scala model implement it's own source-jar navigation policy, or should the Scala importer somehow populate the Java model?

Thanks for your help!

Reply:
@brendandouglas thank you so much for your insights so far. If you'll be able to give us a bit more guidance before the holiday season that would really help us :) Thanks!